### PR TITLE
DBZ-8073: Added timeout for replication slot creation command

### PR DESCRIPTION
**Pull request by rajdangwal**
Addressing: https://issues.redhat.com/browse/DBZ-8073

### CODE REVIEW

**Comment by jpechane** [comment]
Failure is in the new test

**Comment by rajdangwal** [comment]
The tests were failing due to same slot name: `ERROR:  replication slot "test1" already exists` thus failing the assertions. Renamed the slots used in the tests.

**Comment by rajdangwal** [comment]
The failure seems to be on an unrelated test-case and it was successful in the past runs.

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java [hunk 1]
```diff
@@ -521,13 +521,47 @@ public Optional<SlotCreationResult> createReplicationSlot() throws SQLException
...
+            final Duration delay = connectorConfig.retryDelay();
+            int tryCount = 0;
+            while (true) {
+                try {
+                    stmt.execute(createCommand);
+                    break;
+                }
+                catch (SQLException ex) {
+                    // intercept the statement timeout error and retry
+                    if (ex.getMessage().contains("canceling statement due to user request")) {
```
[the remark this answers is not in the document]
Comment (psoo) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:542]: On further reflection, i think we should check for `SQLSTATE 55P03`, too. That is `lock_not_available` and is thrown by PostgreSQL if someone has configured server side `lock_timeout` parameter and if this value is lower than the one configured in Debezium, this code can handle this the same way IMHO.

**Comment by jpechane** [comment]
I tried to run both the failing test method and test class locally and had no issues, so I think it is fine.

**Comment by jpechane** [comment]
@rajdangwal Applied, thanks!

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:542]: @rajdangwal This is definitely good point, @rajdangwal could you please log a follow-up Jira for that as there is one more instance of the same approach in the same class?

@psoo Is the error code unique? I mean is it possible that there are multiple error (error messages) thrown with a the same error code?

**Comment by psoo** [comment]
> @rajdangwal Applied, thanks!

Can we please revisit the exception condition handling per my review comments above? I really dislike acting on errors by parsing possible localized error message. This is very fragile.

**Comment by jpechane** [comment]
@psoo Please see https://github.com/debezium/debezium/pull/5747#discussion_r1708620287

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (psoo) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:542]: @jpechane Sorry, saw your comment too late.

The "problem" is indeed the `query_canceled` SQLSTATE. This can also be our backend canceled by a `SIGINT` or `pg_cancel_backend()` by the DBA. I assume pgjdbc calls `cancel()` on the statement handle after receiving the timeout, which does the same. So we can't distinguish whether the DBA cancelled our session or we ran into a timeout.

`lock_not_available` in contrast should be pretty unique: it means the backend couldn't gain the requested lock and timed out.

That said, SQLSTATE is usually pretty well defined what conditions they describe. And i don't think that `query_canceled` is a problem here: we're already trying `MaxRetries` and then giving up. If there is a real problem and the DBA wants to get us out of the way, he should call `pg_terminate_backend()` or SIGTERM us, which gives a `admin_shutdown` (57P01) SQLSTATE.

In summary, i think this is much better than parsing the error message: localized instances would just render our retry logic here useless.

[the remark this answers is not in the document]
Comment (rajdangwal) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:542]: Thanks @psoo @jpechane 
Sure, I will log a follow-up Jira for handling the exceptions based on `SQLSTATE` instead of localized error messages.

[the remark this answers is not in the document]
Comment (rajdangwal) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:542]: Created: https://issues.redhat.com/browse/DBZ-8127

Will follow-up with a PR.

[the remark this answers is not in the document]
Comment (rajdangwal) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java:542]: Raised a PR here: https://github.com/debezium/debezium/pull/5763
Please take a look. Thanks.
